File: go/xwebhook.go
package xwebhook

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/x-webhook/x-webhooks/go/internal/openapi"
	"github.com/x-webhook/x-webhooks/go/internal/version"
)

type (
	XwebhookOptions struct {
		Debug bool

		// Overrides the base URL (protocol + hostname) used for all requests sent by this Xwebhook client. (Useful for testing)
		ServerUrl  *url.URL
		HTTPClient *http.Client
	}
	Xwebhook struct {
		Authentication *Authentication
		Application    *Application
		Endpoint       *Endpoint
		EventType      *EventType
		Integration    *Integration
		Message        *Message
		MessageAttempt *MessageAttempt
		Statistics     *Statistics
	}
)

var defaultHTTPClient = &http.Client{
	Timeout: 60 * time.Second,
}

func New(token string, options *XwebhookOptions) *Xwebhook {
	conf := openapi.NewConfiguration()
	conf.Scheme = "https"
	conf.Host = "api.webhook.openweb3.io"
	conf.HTTPClient = defaultHTTPClient

	var tokenParts = strings.Split(token, ".")
	var region = tokenParts[len(tokenParts)-1]
	if region == "us" {
		conf.Host = "api.us.webhook.openweb3.io"
	} else if region == "eu" {
		conf.Host = "api.eu.webhook.openweb3.io"
	} else if region == "in" {
		conf.Host = "api.in.webhook.openweb3.io"
	}

	if options != nil {
		conf.Debug = options.Debug
		if options.ServerUrl != nil {
			conf.Scheme = options.ServerUrl.Scheme
			conf.Host = options.ServerUrl.Host
		}
		if options.HTTPClient != nil {
			conf.HTTPClient = options.HTTPClient
		}
	}
	conf.AddDefaultHeader("Authorization", fmt.Sprintf("Bearer %s", token))
	conf.UserAgent = fmt.Sprintf("xwebhook-libs/%s/go", version.Version)
	apiClient := openapi.NewAPIClient(conf)
	return &Xwebhook{
		Authentication: &Authentication{
			api: apiClient,
		},
		Application: &Application{
			api: apiClient,
		},
		Endpoint: &Endpoint{
			api: apiClient,
		},
		EventType: &EventType{
			api: apiClient,
		},
		Message: &Message{
			api: apiClient,
		},
		Integration: &Integration{
			api: apiClient,
		},
		MessageAttempt: &MessageAttempt{
			api: apiClient,
		},
		Statistics: &Statistics{
			api: apiClient,
		},
	}
}

File: java/lib/src/main/java/io/openweb3/xwebhook/Xwebhook.java
package io.openweb3.xwebhook;

import io.openweb3.xwebhook.internal.ApiClient;
import io.openweb3.xwebhook.internal.Configuration;
import io.openweb3.xwebhook.internal.auth.HttpBearerAuth;

public final class Xwebhook {
	public static final String VERSION = "1.24.0";
	private final Application application;
	private final Authentication authentication;
	private final Endpoint endpoint;
	private final EventType eventType;
	private final Integration integration;
	private final Message message;
	private final MessageAttempt messageAttempt;
	private final Statistics statistics;

	public Xwebhook(final String token) {
		this(token, new XwebhookOptions());
	}

	public Xwebhook(final String token, final XwebhookOptions options) {
		ApiClient apiClient = new ApiClient();

		String[] tokenParts = token.split("\\.");
		String region = tokenParts[tokenParts.length - 1];
		if (region.equals("us")) {
			apiClient.setBasePath("https://api.webhook.openweb3.io");
		} else if (region.equals("eu")) {
			apiClient.setBasePath("https://api.webhook.openweb3.io");
		} else if (region.equals("in")) {
			apiClient.setBasePath("https://api.webhook.openweb3.io");
		} else {
			apiClient.setBasePath(options.getServerUrl());
		}

		apiClient.setUserAgent(String.format("xwebhook-libs/%s/java", Xwebhook.VERSION));

		HttpBearerAuth httpBearer = (HttpBearerAuth) apiClient.getAuthentication("HTTPBearer");
		httpBearer.setBearerToken(token);

		Configuration.setDefaultApiClient(apiClient);

		application = new Application();
		authentication = new Authentication();
		endpoint = new Endpoint();
		eventType = new EventType();
		integration = new Integration();
		message = new Message();
		messageAttempt = new MessageAttempt();
		statistics = new Statistics();
	}

	public Application getApplication() {
		return application;
	}

	public Authentication getAuthentication() {
		return authentication;
	}

	public Endpoint getEndpoint() {
		return endpoint;
	}

	public EventType getEventType() {
		return eventType;
	}

	public Integration getIntegration() {
		return integration;
	}

	public Message getMessage() {
		return message;
	}

	public MessageAttempt getMessageAttempt() {
		return messageAttempt;
	}

	public Statistics getStatistics() {
		return statistics;
	}
}
